[virt] Partial fix for issue #57

arodchen opened the pull request.

arodchen committed:
Merge pull request #1 from s5z/master

Sync with s5z/zsim

arodchen committed:
[virt] SyscallEnter is split into two methods SyscallEnterUnlocked and SyscallEnterLocked to avoid waitUntilQueued time-outs due to the pin internal lock
The functionality of SyscallEnter is moved to SyscallEnterUnlocked called without acquiring the Pin internal lock. The actual modification of system call arguments is done in SyscallEnterLocked with acquiring the Pin internal lock. This commit fixes the following issue: https://github.com/s5z/zsim/issues/57

arodchen committed:
[virt] Fixed SaveModifiedSyscallArgs call site.

arodchen committed:
[virt] Retrying timed out system calls for the remainder of sleep time when transitioning to FF

arodchen committed:
[virt] Fine-grain locking for Scheduler::isSleeping.
Prevents holding the Pin internal lock in PostPatchTimeoutSyscall, when another thread is in waitUntilQueued.

arodchen committed:
Merge pull request #2 from s5z/master

Sync with s5z/zsim

arodchen committed:
Merge pull request #3 from s5z/master

Merge pull request #3 from s5z/master

gaomy3832 commented:
I think this pull request is not necessary any more. Please see my comment in #57.

arodchen commented:
I think it fixes some of the problems. My comments are in #57. 
It also contains fixes unrelated to #57 for the following issues:
1. Deadlock at Scheduler::isSleeping() check.
2. Retrying timed out system calls for the remainder of sleep time when transitioning to FF mode.

arodchen committed:
[virt] Fixed timeout virtualization for FUTEX_CLOCK_REALTIME flag.
If FUTEX_CLOCK_REALTIME flag is set, the timeout should be treated as absolute based on CLOCK_REALTIME.

arodchen committed:
[scheduler] Added DEBUG_SCHEDULER macros for debugging purposes and all commented calls to info(...) were substututed.

arodchen committed:
[scheduler] Descheduling finishing thread on the condition that it has been scheduled.

arodchen committed:
[ooo_core] No store forwarding for false-predicated loads.

arodchen committed:
[build] Fixed syscalls identification on 64-bit Ubuntu. If /usr/include/asm/unistd.h does not exist, processing /usr/include/x86_64-linux-gnu/asm/unistd.h instead.

arodchen committed:
[scheduler] Supported finishing of a sleeping thread.
